### DISCUSSION TITLE
multiple propagation headers: B3 and W3C

### DISCUSSION THREAD

**Discussion by ffatghub** [open-telemetry/opentelemetry-java#5000] [Q&A]
Hi all, I am new to otel please bear with me,

in case of manual instrumentation is it possible to use 2 propagators for getting (extract) and setting (inject) both B3 and W3C headers? Something similar to - if I understand it right - what can be achieved with "-Dotel.propagators=tracecontext,baggage,b3multi" and javaagent.

thanks a lot

**Comment by jkwatson** [comment]
See the docs here for how to assign multiple propagators: https://opentelemetry.io/docs/instrumentation/java/manual/#setup

**Comment by ffatghub** [comment]
> See the docs here for how to assign multiple propagators: https://opentelemetry.io/docs/instrumentation/java/manual/#setup

Hi @jkwatson , thanks. You mean something like this?

```
        return OpenTelemetrySdk.builder()
                .setTracerProvider(sdkTracerProvider)
                .setMeterProvider(sdkMeterProvider)
                .setPropagators(ContextPropagators.create(B3Propagator.injectingMultiHeaders()))
                .setPropagators(ContextPropagators.create(W3CTraceContextPropagator.getInstance()))
                .buildAndRegisterGlobal();
```

How OTEL decide which propagator to use for outbound HTTP requests (inject time)?

thank you

**Comment by jkwatson** [comment]
> > See the docs here for how to assign multiple propagators: https://opentelemetry.io/docs/instrumentation/java/manual/#setup
> 
> Hi @jkwatson , thanks. You mean something like this?
> 
> ```
>         return OpenTelemetrySdk.builder()
>                 .setTracerProvider(sdkTracerProvider)
>                 .setMeterProvider(sdkMeterProvider)
>                 .setPropagators(ContextPropagators.create(B3Propagator.injectingMultiHeaders()))
>                 .setPropagators(ContextPropagators.create(W3CTraceContextPropagator.getInstance()))
>                 .buildAndRegisterGlobal();
> ```
> 
> How OTEL decide which propagator to use for outbound HTTP requests (inject time)?
> 
> thank you
> 
> 

Context propagators.create(...) takes vararg parameters. 

It will inject all of the headers from all of the configured propagators.

**Comment by ffatghub** [comment]
> W3CTraceContextPropagator.getInstance()

Hi @jkwatson , create() in opentelemetry-sdk.1.19.0.jar takes - if I had enough coffee :-) - one argument.

```
static ContextPropagators create(TextMapPropagator textPropagator) {
        Objects.requireNonNull(textPropagator, "textPropagator");
        return new DefaultContextPropagators(textPropagator);
}
```

ciao thanks

**Comment by jkwatson** [comment]
> > W3CTraceContextPropagator.getInstance()
> 
> 
> 
> Hi @jkwatson , create() in opentelemetry-sdk.1.19.0.jar takes - if I had enough coffee :-) - one argument.
> 
> 
> 
> ```
> 
> static ContextPropagators create(TextMapPropagator textPropagator) {
> 
>         Objects.requireNonNull(textPropagator, "textPropagator");
> 
>         return new DefaultContextPropagators(textPropagator);
> 
> }
> 
> ```
> 
> 
> 
> ciao thanks
> 
> 

Sorry, I misremembered. It's TextMapPropagator.composite(...)

**Comment by ffatghub** [comment]
```
        return OpenTelemetrySdk.builder()
                .setTracerProvider(sdkTracerProvider)
                .setMeterProvider(sdkMeterProvider)
                .setPropagators(ContextPropagators.create(
                        TextMapPropagator.composite(
                                B3Propagator.injectingMultiHeaders(),
                                W3CTraceContextPropagator.getInstance())))
                .buildAndRegisterGlobal();
```

thanks

**Comment by jkwatson** [comment]
I'm going to turn this into a discussion, since it's not an issue.
